### PR TITLE
Respect model overrides of get_FIELD_display for choices (swev-id: django__django-11999)

### LINKED ISSUE #338: Cannot override get_FIELD_display() in Django 2.2+ (regression fix)
## User Report

Cannot override get_FOO_display() in Django 2.2+.

Description

I cannot override the get_FIELD_display function on models since version 2.2. It works in version 2.1.

Example:

```python
class FooBar(models.Model):
    foo_bar = models.CharField(_("foo"), choices=[(1, 'foo'), (2, 'bar')])

    def __str__(self):
        return self.get_foo_bar_display()  # This returns 'foo' or 'bar' in 2.2, but 'something' in 2.1

    def get_foo_bar_display(self):
        return "something"
```

What I expect is that I should be able to override this function.

Issue Comments:

- Thanks for this report. Regression in a68ea231012434b522ce45c513d84add516afa60. Reproduced at 54a7b021125d23a248e70ba17bf8b10bc8619234.
- OK, I have a lead on this. Not at all happy about how it looks at first pass, but I'll a proof of concept PR together for it tomorrow AM.
- I don't think it should be marked as blocker since it looks like it was never supported, because it depends on the order of attrs passed in ModelBase.__new__(). So on Django 2.1 and Python 3.7:

```
In [1]: import django ...: django.VERSION
In [2]: from django.db import models ...:
   ...: class FooBar(models.Model):
   ...:     def get_foo_bar_display(self):
   ...:         return "something"
   ...:     foo_bar = models.CharField("foo", choices=[(1, 'foo'), (2, 'bar')])
   ...:     def __str__(self):
   ...:         return self.get_foo_bar_display()
   ...:     class Meta:
   ...:         app_label = 'test'
   ...: FooBar(foo_bar=1)
Out[2]: <FooBar: foo>
```

Before Python 3.6 the order of attrs wasn't defined at all.

- Sergey, an example from the ticket description works for me with Django 2.1 and Python 3.6, 3.7 and 3.8.

```
In [2]: import django ...: django.VERSION
Out[2]: (2, 1, 13, 'final', 0)
In [3]: import sys ...: sys.version
Out[3]: '3.5.7 (default, Oct 17 2019, 07:04:41) \n[GCC 8.3.0]'
In [4]: from django.db import models ...:
   ...: class FooBar(models.Model):
   ...:     foo_bar = models.CharField("foo", choices=[(1, 'foo'), (2, 'bar')])
   ...:     def __str__(self):
   ...:         return self.get_foo_bar_display()
   ...:     def get_foo_bar_display(self):
   ...:         return "something"
   ...:     class Meta:
   ...:         app_label = 'test'
   ...: FooBar(foo_bar=1)
Out[4]: <FooBar: foo>
```

- OK, so there is a behaviour change here, but Sergey is correct that it does depend on attr order, so it's hard to say that this can be said to ever have been thought of as supported, with the exact example provided.

This example produces the opposite result on 2.1 (even on >=PY36):

```python
def test_overriding_display_backwards(self):
    class FooBar2(models.Model):
        def get_foo_bar_display(self):
            return "something"
        foo_bar = models.CharField("foo", choices=[(1, 'foo'), (2, 'bar')])
    f = FooBar2(foo_bar=1)
    # This returns 'foo' or 'bar' in both 2.2 and 2.1
    self.assertEqual(f.get_foo_bar_display(), "foo")
```

Because get_foo_bar_display() is defined before foo_bar it gets replaced in the add_to_class() step. Semantically order shouldn't make a difference. Given that it does, I can't see that we're bound to maintain that behaviour. (There's a possible fix in Field.contribute_to_class() but implementing that just reverses the pass/fail behaviour depending on order...)

Rather, the correct way to implement this on 2.2+ is:

```python
def test_overriding_display(self):
    class FooBar(models.Model):
        foo_ bar = models.CharField("foo", choices=[(1, 'foo'), (2, 'bar')])
        def _get_FIELD_display(self, field):
            if field.attname == 'foo_bar':
                return "something"
            return super()._get_FIELD_display(field)
    f = FooBar(foo_bar=1)
    self.assertEqual(f.get_foo_bar_display(), "something")
```

This is stable for declaration order on version 2.2+. This approach requires overriding _get_FIELD_display() before declaring fields on 2.1, because otherwise Model.get_FIELD_display() is picked up during Field.contribute_to_class(). This ordering dependency is the same issue that was addressed in a68ea231012434b522ce45c513d84add516afa60, and the follow-up in #30254. The behaviour in 2.1 (and before) was incorrect.

Yes, there's a behaviour change here but it's a bugfix, and all bugfixes are breaking changes if you're depending on the broken behaviour. I'm going to downgrade this from Release Blocker accordingly. I'll reclassify this as a Documentation issue and provide the working example, as overriding get_FIELD_display() is a legitimate use-case.

Replying to Carlton Gibson:

> (There's a possible fix in Field.contribute_to_class() but implementing that just reverses the pass/fail behaviour depending on order...)

Doesn't this fix it?

```python
if not hasattr(cls, 'get%s_display' % self.name):
    setattr(cls, 'get_%s_display' % self.name, partialmethod(cls._get_FIELD_display, field=self))
```

---

## Research Specification

- Current injection:
  - django/db/models/fields/__init__.py::Field.contribute_to_class() unconditionally sets `get_<field>_display` if `choices` is not None, using `partialmethod(cls._get_FIELD_display, field=self)`.
- Proposed fix:
  - Guard the injection to respect user-defined overrides: only inject if the class does not already define `get_<field>_display`.
  - Implementation guideline:
    - Replace unconditional `setattr()` with:
      - `if self.choices is not None and not hasattr(cls, f'get_{self.name}_display'):`
      - then perform the existing setattr with `partialmethod`.
- Tests to add:
  1. Override method declared before field: ensure user method is used.
  2. Override method declared after field: ensure user method is used.
  3. Override `_get_FIELD_display` conditionally on `field.attname`: ensure injected partialmethod respects this.
  4. Default choices behavior remains intact (labels, fallback to raw value, None).
  5. Nested choices/translation remain intact.
- Files to modify:
  - Code: `django/db/models/fields/__init__.py` in `Field.contribute_to_class()`.
  - Tests: `tests/model_fields/test_display_overrides.py` (new) or alongside existing tests.
- Edge cases:
  - Proxy models, multiple inheritance: covered by method resolution; guard avoids overwriting overrides.
  - Deferred fields: no behavior change expected; out of scope.
  - Name vs attname vs db_column: unchanged; public API uses field.name.

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- guard `Field.contribute_to_class()` so existing `get_<field>_display()` overrides survive choice injection
- add regression tests covering overrides declared before/after the field, custom `_get_FIELD_display()`, and default display fallbacks

## Reproduction
1. Add `tests/model_fields/test_display_overrides_tmp.py` with:
   ```python
   from django.db import models
   from django.test import SimpleTestCase


   class DisplayOverrideFailureTests(SimpleTestCase):
       def test_override_after_field_fails_currently(self):
           class Hat(models.Model):
               STYLE_CHOICES = [("B", "Bowler"), ("F", "Fedora")]
               style = models.CharField(max_length=1, choices=STYLE_CHOICES)

               def get_style_display(self):
                   return f"custom:{self.style}"

               class Meta:
                   app_label = "tests"

           self.assertEqual(Hat(style="B").get_style_display(), "custom:B")
   ```
2. Run `/workspace/.venv/bin/python tests/runtests.py model_fields.test_display_overrides_tmp --parallel=1`

```
F
======================================================================
FAIL: test_override_after_field_fails_currently (model_fields.test_display_overrides_tmp.DisplayOverrideFailureTests.test_override_after_field_fails_currently)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/workspace/django/tests/model_fields/test_display_overrides_tmp.py", line 17, in test_override_after_field_fails_currently
    self.assertEqual(Hat(style="B").get_style_display(), "custom:B")
AssertionError: 'Bowler' != 'custom:B'
- Bowler
+ custom:B
```

## Testing
- `/workspace/.venv/bin/python tests/runtests.py model_fields.test_display_overrides --parallel=1`
- `/workspace/.venv/bin/python tests/runtests.py model_fields --parallel=1`

Fixes #338

### CODE REVIEW

**Comment by casey-brooks** [comment]
## Local Validation
- `/workspace/.venv/bin/flake8 django/db/models/fields/__init__.py tests/model_fields/test_display_overrides.py` *(no issues)*
- `/workspace/.venv/bin/python tests/runtests.py model_fields.test_display_overrides --parallel=1` *(4 tests, 0 failed)*
- `/workspace/.venv/bin/python tests/runtests.py model_fields --parallel=1` *(314 tests, 0 failed, 2 skipped)*

File: django/db/models/fields/__init__.py [hunk 1]
```diff
@@ -762,8 +762,9 @@ def contribute_to_class(self, cls, name, private_only=False):
             # such fields can't be deferred (we don't have a check for this).
             if not getattr(cls, self.attname, None):
                 setattr(cls, self.attname, self.descriptor_class(self))
-        if self.choices is not None:
-            setattr(cls, 'get_%s_display' % self.name,
+        display_method_name = 'get_%s_display' % self.name
+        if self.choices is not None and not hasattr(cls, display_method_name):
```
Comment (noa-lucent) [inline django/db/models/fields/__init__.py]: [major] Guarding on `hasattr(cls, display_method_name)` prevents us from re-installing the generated display accessor when a subclass replaces a field defined on an abstract base class with a different set of `choices`. In that scenario the subclass still inherits the auto-generated `partialmethod` from the base class, so this condition short-circuits and the subclass will continue to call the base method, which uses the base field's choices. For example:

```python
class Base(models.Model):
    class Meta:
        abstract = True
    status = models.CharField(max_length=1, choices=[("A", "Alpha")])

class Child(Base):
    status = models.CharField(max_length=1, choices=[("B", "Beta")])
```

`Child(status="B").get_status_display()` returns "Alpha" after this change because `hasattr()` sees the inherited accessor and we never re-bind it to the new field.

We need to skip the auto-generation only when there is a user override, but still refresh the generated accessor when the existing attribute is one of the auto-generated `partialmethod`s. Could you tighten the predicate (e.g. inspect `getattr(cls, display_method_name)` and only bail out when it is *not* a `functools.partialmethod`) and add a regression test covering an abstract base class whose concrete subclass overrides the `choices`?

File: tests/model_fields/test_display_overrides.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            def _get_FIELD_display(self, field):
+                value = super()._get_FIELD_display(field)
+                return f"custom:{value}"
+
+        self.assertEqual(Jacket(style="T").get_style_display(), "custom:Trench")
+
+    def test_get_field_display_default_choices_label(self):
+        class Scarf(models.Model):
+            MATERIAL_CHOICES = [("C", "Cotton")]
+            material = models.CharField(max_length=1, choices=MATERIAL_CHOICES, null=True)
```
Comment (noa-lucent) [inline tests/model_fields/test_display_overrides.py]: [nit] Please wrap this line to keep it within Django's 79-character limit (including indentation).

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Major regression risk: the new guard blocks refreshing the generated display accessor when subclasses override an abstract base field's choices; see inline comment. Also noted a small style nit.

**Comment by casey-brooks** [comment]
## Local Validation (update)
- `/workspace/.venv/bin/flake8 django/db/models/fields/__init__.py tests/model_fields/test_display_overrides.py` *(no issues)*
- `/workspace/.venv/bin/python tests/runtests.py model_fields.test_display_overrides --parallel=1` *(5 tests, 0 failed)*
- `/workspace/.venv/bin/python tests/runtests.py model_fields --parallel=1` *(315 tests, 0 failed, 2 skipped)*

**Review by noa-lucent — APPROVED** [review]
Thanks for the quick follow-up. Rechecked the guard and it now only regenerates the accessor when the inherited attribute is Django's auto-installed partialmethod, so user overrides still win while subclasses that replace the field rebind correctly. The new regression test around abstract base fields verifies the coverage gap I was worried about, and the existing tests continue to cover both declaration orders and the custom `_get_FIELD_display()` path. Looks good to me.
